### PR TITLE
RFC for caching results of `treeFor` hook

### PR DESCRIPTION
**Pull request by rwjblue**
[Rendered](https://github.com/ember-cli/rfcs/blob/addon-tree-caching/active/0000-addon-tree-caching.md)

[Implementation PR](https://github.com/ember-cli/ember-cli/pull/6573) here.

### CODE REVIEW

File: active/0000-addon-tree-caching.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+- ember-cli's custom [`mergeTrees` implementation](https://github.com/ember-cli/ember-cli/blob/4ec7b5951e8a9dd292029faf20d1858abf7bdfa0/lib/broccoli/merge-trees.js) (which is already aware of other tree reduction techniques) will be updated so that calling `mergeTrees([treeA, treeA]);` simply returns `treeA`, and `mergeTrees([treeA, treeB, treeA])` removes the duplicated `treeA` in the input nodes.
+ 
+The proposed declaration for `Addon.prototype.cacheKeyForTree` in Typescript syntax is:
+
+``` ts
+function cacheKeyForTree(treeType: string): string;
+```
+The default implementation for `Addon.prototype.cacheKeyForTree` will:
+
+- Utilize a shared NPM package (e.g. `calculate-cache-key-for-tree`) that will generate a cache key that incorporates at least the following pieces of information:
```
Comment (stefanpenner) [inline active/0000-addon-tree-caching.md:39]: An additional or complementary heuristic may be that they are at the exact same place on disk?

**Review by stefanpenner — APPROVED** [review]

File: active/0000-addon-tree-caching.md (hunk 1 above)
Comment (rwjblue) [reply active/0000-addon-tree-caching.md:39]: @stefanpenner - Ya, I'm happy to add that also.

Comment (stefanpenner) [reply active/0000-addon-tree-caching.md:39]: @rwjblue only if you think it adds value. Just wanted to share, in-case it was overlooked.

**Review by hjdivad — APPROVED** [review]
LGTM

**Review by trentmwillis — APPROVED** [review]
Looks good to me. Only potential suggestion would be mentioning [the cache API](https://github.com/ember-cli/ember-cli/pull/6589) and maybe link to [cache-key-for-tree](https://github.com/ember-cli/calculate-cache-key-for-tree).

**Comment by kellyselden** [comment]
This was promoted to Final Comment Period in our team meeting.

**Comment by rwjblue** [comment]
Discussed at ember-cli core team meeting today. No new information was brought up during the final comment period, and all of the ember-cli core team is in favor.

Merging...

**Pull request merged**